# <b><u> Project Title : Sales Prediction : Predicting sales of a major store chain Rossmann</u></b>

## <b> Problem Description </b>

### Rossmann operates over 3,000 drug stores in 7 European countries. Currently, Rossmann store managers are tasked with predicting their daily sales for up to six weeks in advance. Store sales are influenced by many factors, including promotions, competition, school and state holidays, seasonality, and locality. With thousands of individual managers predicting sales based on their unique circumstances, the accuracy of results can be quite varied.

### You are provided with historical sales data for 1,115 Rossmann stores. The task is to forecast the "Sales" column for the test set. Note that some stores in the dataset were temporarily closed for refurbishment.

## <b> Data Description </b>

### <b>Rossmann Stores Data.csv </b> - historical data including Sales
### <b>store.csv </b> - supplemental information about the stores


### <b><u>Data fields</u></b>
### Most of the fields are self-explanatory. The following are descriptions for those that aren't.

* #### Id - an Id that represents a (Store, Date) duple within the test set
* #### Store - a unique Id for each store
* #### Sales - the turnover for any given day (this is what you are predicting)
* #### Customers - the number of customers on a given day
* #### Open - an indicator for whether the store was open: 0 = closed, 1 = open
* #### StateHoliday - indicates a state holiday. Normally all stores, with few exceptions, are closed on state holidays. Note that all schools are closed on public holidays and weekends. a = public holiday, b = Easter holiday, c = Christmas, 0 = None
* #### SchoolHoliday - indicates if the (Store, Date) was affected by the closure of public schools
* #### StoreType - differentiates between 4 different store models: a, b, c, d
* #### Assortment - describes an assortment level: a = basic, b = extra, c = extended
* #### CompetitionDistance - distance in meters to the nearest competitor store
* #### CompetitionOpenSince[Month/Year] - gives the approximate year and month of the time the nearest competitor was opened
* #### Promo - indicates whether a store is running a promo on that day
* #### Promo2 - Promo2 is a continuing and consecutive promotion for some stores: 0 = store is not participating, 1 = store is participating
* #### Promo2Since[Year/Week] - describes the year and calendar week when the store started participating in Promo2
* #### PromoInterval - describes the consecutive intervals Promo2 is started, naming the months the promotion is started anew. E.g. "Feb,May,Aug,Nov" means each round starts in February, May, August, November of any given year for that store

In [1]:
#important libraries imported
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats 
from datetime import date
import warnings
warnings.filterwarnings('ignore')

from numpy import math
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from scipy.stats import zscore
import missingno as msno
import xgboost as xg
from sklearn.linear_model import SGDRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import Ridge, RidgeCV
from sklearn.linear_model import Lasso, LassoCV


In [2]:
#mounting the drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
#loading the dataset
dataset_store=pd.read_csv('/content/drive/MyDrive/EDA/SL_Regr/store.csv')
dataset_sales_data=pd.read_csv('/content/drive/MyDrive/EDA/SL_Regr/Rossmann_Stores_Data.csv')

Firstly we will look at both the datasets to understand the variables involved and try to get the first impression about the features and target variables.

# Data Reading 

In [4]:
#dataset reading
dataset_store.head()

Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      2         a          a                570.0                       11.0   
2      3         a          a              14130.0                       12.0   
3      4         c          c                620.0                        9.0   
4      5         a          a              29910.0                        4.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2007.0       1             13.0           2010.0   
2                    2006.0       1             14.0           2011.0   
3                    2009.0       0              NaN              NaN   
4                    2015.0       0              NaN              NaN   

     PromoInterval  
0              NaN  
1  Jan,Apr,Jul,Oct  
2  Jan,Apr,Jul,Oct  
3              NaN  
4              NaN

In [5]:
dataset_store.tail()

Store StoreType Assortment  CompetitionDistance  \
1110   1111         a          a               1900.0   
1111   1112         c          c               1880.0   
1112   1113         a          c               9260.0   
1113   1114         a          c                870.0   
1114   1115         d          c               5350.0   

      CompetitionOpenSinceMonth  CompetitionOpenSinceYear  Promo2  \
1110                        6.0                    2014.0       1   
1111                        4.0                    2006.0       0   
1112                        NaN                       NaN       0   
1113                        NaN                       NaN       0   
1114                        NaN                       NaN       1   

      Promo2SinceWeek  Promo2SinceYear     PromoInterval  
1110             31.0           2013.0   Jan,Apr,Jul,Oct  
1111              NaN              NaN               NaN  
1112              NaN              NaN               NaN  
1113              NaN              NaN               NaN  
1114             22.0           2012.0  Mar,Jun,Sept,Dec

In [6]:
dataset_sales_data.head()

Store  DayOfWeek        Date  Sales  Customers  Open  Promo StateHoliday  \
0      1          5  2015-07-31   5263        555     1      1            0   
1      2          5  2015-07-31   6064        625     1      1            0   
2      3          5  2015-07-31   8314        821     1      1            0   
3      4          5  2015-07-31  13995       1498     1      1            0   
4      5          5  2015-07-31   4822        559     1      1            0   

   SchoolHoliday  
0              1  
1              1  
2              1  
3              1  
4              1

In [7]:
dataset_sales_data.tail()

Store  DayOfWeek        Date  Sales  Customers  Open  Promo  \
1017204   1111          2  2013-01-01      0          0     0      0   
1017205   1112          2  2013-01-01      0          0     0      0   
1017206   1113          2  2013-01-01      0          0     0      0   
1017207   1114          2  2013-01-01      0          0     0      0   
1017208   1115          2  2013-01-01      0          0     0      0   

        StateHoliday  SchoolHoliday  
1017204            a              1  
1017205            a              1  
1017206            a              1  
1017207            a              1  
1017208            a              1

In [8]:
# Length of the historical sales dataset
dataset_sales_data.Store.value_counts()
dataset_sales_data.shape

(1017209, 9)

As we can see above there are two datasets with a store IDs as a common column for both of them.
So before starting evaluating we should merge the datasets on Store column. 

In [9]:
#merging the two datasets
Dataset=pd.merge(dataset_store,dataset_sales_data,on='Store')
Dataset.head(5)


Store StoreType Assortment  CompetitionDistance  CompetitionOpenSinceMonth  \
0      1         c          a               1270.0                        9.0   
1      1         c          a               1270.0                        9.0   
2      1         c          a               1270.0                        9.0   
3      1         c          a               1270.0                        9.0   
4      1         c          a               1270.0                        9.0   

   CompetitionOpenSinceYear  Promo2  Promo2SinceWeek  Promo2SinceYear  \
0                    2008.0       0              NaN              NaN   
1                    2008.0       0              NaN              NaN   
2                    2008.0       0              NaN              NaN   
3                    2008.0       0              NaN              NaN   
4                    2008.0       0              NaN              NaN   

  PromoInterval  DayOfWeek        Date  Sales  Customers  Open  Promo  \
0           NaN          5  2015-07-31   5263        555     1      1   
1           NaN          4  2015-07-30   5020        546     1      1   
2           NaN          3  2015-07-29   4782        523     1      1   
3           NaN          2  2015-07-28   5011        560     1      1   
4           NaN          1  2015-07-27   6102        612     1      1   

  StateHoliday  SchoolHoliday  
0            0              1  
1            0              1  
2            0              1  
3            0              1  
4            0              1

Next we will look for description and summaries of dataset to see number of  missing values and to get the general idea about the distribution of values and the data types of the variables involved.

In [10]:
# description and summary of the variables involved
Dataset.describe(include='all')

Store StoreType Assortment  CompetitionDistance  \
count   1.017209e+06   1017209    1017209         1.014567e+06   
unique           NaN         4          3                  NaN   
top              NaN         a          a                  NaN   
freq             NaN    551627     537445                  NaN   
mean    5.584297e+02       NaN        NaN         5.430086e+03   
std     3.219087e+02       NaN        NaN         7.715324e+03   
min     1.000000e+00       NaN        NaN         2.000000e+01   
25%     2.800000e+02       NaN        NaN         7.100000e+02   
50%     5.580000e+02       NaN        NaN         2.330000e+03   
75%     8.380000e+02       NaN        NaN         6.890000e+03   
max     1.115000e+03       NaN        NaN         7.586000e+04   

        CompetitionOpenSinceMonth  CompetitionOpenSinceYear        Promo2  \
count               693861.000000             693861.000000  1.017209e+06   
unique                        NaN                       NaN           NaN   
top                           NaN                       NaN           NaN   
freq                          NaN                       NaN           NaN   
mean                     7.222866               2008.690228  5.005638e-01   
std                      3.211832                  5.992644  4.999999e-01   
min                      1.000000               1900.000000  0.000000e+00   
25%                      4.000000               2006.000000  0.000000e+00   
50%                      8.000000               2010.000000  1.000000e+00   
75%                     10.000000               2013.000000  1.000000e+00   
max                     12.000000               2015.000000  1.000000e+00   

        Promo2SinceWeek  Promo2SinceYear    PromoInterval     DayOfWeek  \
count     509178.000000    509178.000000           509178  1.017209e+06   
unique              NaN              NaN                3           NaN   
top                 NaN              NaN  Jan,Apr,Jul,Oct           NaN   
freq                NaN              NaN           293122           NaN   
mean          23.269093      2011.752774              NaN  3.998341e+00   
std           14.095973         1.662870              NaN  1.997391e+00   
min            1.000000      2009.000000              NaN  1.000000e+00   
25%           13.000000      2011.000000              NaN  2.000000e+00   
50%           22.000000      2012.000000              NaN  4.000000e+00   
75%           37.000000      2013.000000              NaN  6.000000e+00   
max           50.000000      2015.000000              NaN  7.000000e+00   

              Date         Sales     Customers          Open         Promo  \
count      1017209  1.017209e+06  1.017209e+06  1.017209e+06  1.017209e+06   
unique         942           NaN           NaN           NaN           NaN   
top     2015-07-31           NaN           NaN           NaN           NaN   
freq          1115           NaN           NaN           NaN           NaN   
mean           NaN  5.773819e+03  6.331459e+02  8.301067e-01  3.815145e-01   
std            NaN  3.849926e+03  4.644117e+02  3.755392e-01  4.857586e-01   
min            NaN  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%            NaN  3.727000e+03  4.050000e+02  1.000000e+00  0.000000e+00   
50%            NaN  5.744000e+03  6.090000e+02  1.000000e+00  0.000000e+00   
75%            NaN  7.856000e+03  8.370000e+02  1.000000e+00  1.000000e+00   
max            NaN  4.155100e+04  7.388000e+03  1.000000e+00  1.000000e+00   

       StateHoliday  SchoolHoliday  
count       1017209   1.017209e+06  
unique            5            NaN  
top               0            NaN  
freq         855087            NaN  
mean            NaN   1.786467e-01  
std             NaN   3.830564e-01  
min             NaN   0.000000e+00  
25%             NaN   0.000000e+00  
50%             NaN   0.000000e+00  
75%             NaN   0.000000e+00  
max             NaN   1.000000e+00

In [11]:
# information of variables involved
Dataset.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1017209 entries, 0 to 1017208
Data columns (total 18 columns):
 #   Column                     Non-Null Count    Dtype  
---  ------                     --------------    -----  
 0   Store                      1017209 non-null  int64  
 1   StoreType                  1017209 non-null  object 
 2   Assortment                 1017209 non-null  object 
 3   CompetitionDistance        1014567 non-null  float64
 4   CompetitionOpenSinceMonth  693861 non-null   float64
 5   CompetitionOpenSinceYear   693861 non-null   float64
 6   Promo2                     1017209 non-null  int64  
 7   Promo2SinceWeek            509178 non-null   float64
 8   Promo2SinceYear            509178 non-null   float64
 9   PromoInterval              509178 non-null   object 
 10  DayOfWeek                  1017209 non-null  int64  
 11  Date                       1017209 non-null  object 
 12  Sales                      1017209 non-null  int64  
 13  Customers   

As we can see that there are 9 Numeric variables and 7 object data type variables for every Store ID and corresponding numeric sales value.
There are 1017209 different sales entries, but there are several Null values in CompetitionOpenSinceMonth/Year columns and PromoSinceWeek/Year/Interval as well.
So further data imputation/ data cleaning is required.